### ISSUE TITLE
Support Impeller on Android

### ISSUE THREAD

**Issue by JgomesAT** [juliansteenbakker/mobile_scanner#867]
_This message has been updated to outline the most up-to-date information on this bug. The original comment is kept at the bottom_

With the release of Flutter 3.27, Impeller on Android has released to stable. This issue tracks adding support for Impeller in mobile_scanner.

We should do the following:
1) Follow the steps from https://docs.flutter.dev/release/breaking-changes/android-surface-plugins _including_ the rotation correction for camera previews
2) Bump the minimum required Flutter version to 3.27, per the note in https://docs.flutter.dev/release/breaking-changes/android-surface-plugins#timeline due to the fact that during the 3.24 - 3.27 timeframe a bug was discovered in the Impeller API for Android, which has been fixed in 3.27
3) Add a TODO to switch from `onSurfaceDestroyed` to `onSurfaceCleanup` in the stable release after Flutter _3.28_
4) Notify users that adding `io.flutter.embedding.android.EnableImpeller` is no longer required for mobile_scanner usage

_The message below is the un-edited original post for this issue_  

With the recent improvement of Impeller on Flutter 3.16 I tried to compile my app with  `<meta-data
            android:name="flutterEmbedding"
            android:value="2" />
            <meta-data
            android:name="io.flutter.embedding.android.EnableImpeller"
            android:value="true" />`
            
  with this, the camera is black and does not work, in iOS with impeller works fine

**Comment by frankmer** [comment]
Same problem for me.

**Comment by navaronbracke** [comment]
Impeller for Android is still in preview. On Android API 29+ this uses Vulkan, which still has some issues with platform views.

**Comment by JgomesAT** [comment]
is there any change with the new Flutter 3.19 and IMpeller, this library is not working with this option

**Comment by navaronbracke** [comment]
As I started above, Impeller on Android is still in preview. Work on external textures has progressed however.
A patch for this is about to land on flutter/engine. If you absolutely need Impeller on Android right now, your best bet is to switch to the master channel. But that is at your own risk, since the master channel is not for production apps.

**Comment by bobekos** [comment]
Hi any news to this? Tested on:
Flutter 3.22 with mobile_scanner 5.1.1 still not working.

**Comment by espresso3389** [comment]
Even official camera/video plugins do not run with Android Impeller (flutter/flutter#139702).

**Comment by JgomesAT** [comment]
I have the same problem, the camera stay in black, I not work in flutter 3.22 with IMpoeller in stable in Android

**Comment by navaronbracke** [comment]
We will have to migrate from `createSurfaceTexture` to `createSurfaceProducer`. This is a new API for Impeller with Vulkan.

**Comment by navaronbracke** [comment]
Small update: there is an upstream regression with the `SurfaceProducer` and Vulkan (OpenGL is unaffected). Will hold off on implementing this until the upstream bug is fixed.

**Comment by rohansohonee1** [comment]
> FYI: Flutter **3.24** has been released.


Do you have any updates to share?

**Comment by navaronbracke** [comment]
I am aware that Flutter 3.24 has landed, and there is an update to the existing Impeller migration doc at https://github.com/flutter/website/pull/11008

I went through that doc before to see what was needed to migrate and I noticed that something was missing (notably the preview orientation, which was the subject of an ongoing discussion on the hackers chat on Discord, which I had followed)

However, I don't think we need to wait for this website doc to land, as what is needed is there for us.
The SurfaceProducer API is available since Flutter 3.22.

I'll update the required steps in the original post in this thread, to reflect the needed steps.

**Comment by navaronbracke** [comment]
I am currently working through several smaller bugs that I want to include in the next release. After that I'll work on the Impeller migration, since I don't want to force our users to switch to 3.24 for the other bugfixes.

**Comment by markuznw** [comment]
Following this, impeller gives a slight performance boost, thanks

**Comment by espresso3389** [comment]
[Flutter 3.27.0](https://medium.com/flutter/whats-new-in-flutter-3-27-28341129570c) stable is released with Android Impeller enabled by default. According to the official release note, we need `<meta-data android:name=”io.flutter.embedding.android.EnableImpeller” android:value=”false” />` on AndroidManifest.xml until mobile_scanner finally supports it.

**Comment by JgomesAT** [comment]
and the schedule to do that?

**Comment by navaronbracke** [comment]
@espresso3389 Thanks for notifying me. I will mark this as P0 in our priority queue and will put this next on the list for the next beta for version 7.0.0

**Comment by navaronbracke** [comment]
This is currently somewhat blocked on an upstream issue. The SurfaceProducer migration is done, but the orientation correction that is mentioned in the migration guide for camera plugins is no longer correct for the current state of Impeller on Android in Flutter 3.27

**Comment by Shakle** [comment]
> This is currently somewhat blocked on an upstream issue. The SurfaceProducer migration is done, but the orientation correction that is mentioned in the migration guide for camera plugins is no longer correct for the current state of Impeller on Android in Flutter 3.27

- They've made some big changes to impeller in 3.29 which was released today, please have a look if it solves current problems. I wait for the new mobile scanner very much, because can't switch our project to impeller without it, thanks.

**Comment by navaronbracke** [comment]
Yes, I will check again with Flutter 3.29 to see if it was resolved.

**Comment by markuznw** [comment]
Seems not yet fixed, PR coming https://github.com/flutter/flutter/issues/154241#issuecomment-2655098620

**Comment by JgomesAT** [comment]
IN my case with the beta 7 now I can see the camera correctly, with Impeller true

**Comment by emakar** [comment]
Hi! Is any help wanted or the issue is being worked on already?

**Comment by navaronbracke** [comment]
@emakar We still need to test / verify that the orientation change of the preview works for older Android devices (so < API 29).
The Impeller migration mainly focused on the preview itself and the rotation for newer devices (API 29+).

But other than that, nothing else is left for this issue.

**Comment by juliansteenbakker** [comment]
I think i've fixed all rotation and aspect ratio issues in #1402. Would be great if someone can verify this. I just need to make some changes to make it working with the apple implementation, and after that i think version 7 will be ready for release

**Comment by juliansteenbakker** [comment]
I have released a release candidate for version 7, v7.0.0-rc.1.
If no issues are found, a stable release will follow next week.

https://github.com/juliansteenbakker/mobile_scanner/releases/tag/v7.0.0-rc.1
